chainermn/optimizers.py

```diff
@@ -46,15 +51,117 @@ def __setattr__(self, attr_name, value):
+            loss.backward()
+            del loss
+
+        if self.is_changed(target, self.target_params_list[0]):
+            self.wait()
+            self.communicator.broadcast_data(target)
+            super(_DoubleBufferingOptimizer, self).__setattr__(
+                'device',
+                chainer.cuda.get_device_from_id(target._device_id))
+            super(_DoubleBufferingOptimizer, self).__setattr__(
+                'communicated_target', copy.deepcopy(target))
+            super(_DoubleBufferingOptimizer, self).__setattr__(
+                'target_params_list', [
+                    list(sorted(self.target.namedparams())),
+                    list(sorted(self.communicated_target.namedparams()))])
+        else:
+            self.wait()
+            self.swap_grad(self.target_params_list[0],
+                           self.target_params_list[1])
+            super(_DoubleBufferingOptimizer, self).__setattr__(
+                'allreduce_grad_res',
+                self.thread_pool.apply_async(self.allreduce_grad))
+            if self.needs_update:
+                self.actual_optimizer.update(None, *args, **kwds)
+            else:
+                super(_DoubleBufferingOptimizer, self).__setattr__(
+                    'needs_update', True)
+
+    def allreduce_grad(self):
+        chainer.cuda.get_device(self.device).use()
+        self.communicator.allreduce_grad(
+            self.communicated_target, self.allreduce_grad_stream)
+
+    def is_changed(self, target, previous_params):
+        target_params = list(sorted(target.namedparams()))
+        if len(previous_params) != len(target_params):
+            return True
+
+        for param1, param2 in zip(target_params, previous_params):
+            name1, var1 = param1
+            name2, var2 = param2
+            if (name1 != name2) or (var1.data is None) != (var2.data is None):
+                return True
+        return False
+
+    def swap_grad(self, target1_params, target2_params):
+        for param1, param2 in zip(target1_params, target2_params):
+            _, var1 = param1
+            _, var2 = param2
+            var1.grad, var2.grad = var2.grad, var1.grad
+
+    def wait(self):
+        if self.allreduce_grad_res is not None:
+            self.allreduce_grad_res.get()
+
+    def __getattr__(self, attr_name):
+        return getattr(self.actual_optimizer, attr_name)
+
+    def __setattr__(self, attr_name, value):
+        setattr(self.actual_optimizer, attr_name, value)
+
+
+def create_multi_node_optimizer(actual_optimizer, communicator,
+                                double_buffering=False):
     """Create a multi node optimizer from a Chainer optimizer.
 
     Args:
         actual_optimizer: Chainer optimizer
             (e.g., ``chainer.optimizers.Adam``).
         communicator: ChainerMN communicator.
-
+        double_buffering: double buffering flag. It is supported
+            by PureNcclCommunicator only.
```

Comment:
What is double buffering? Could you add a paragraph that explains what is double buffering.

Reply:
I suppose you will write a blog post (or something like that) in near future. After it is published, you might want to add URL link to that paragraph.